crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -3462,15 +3462,16 @@ impl<'db> TypeInferenceBuilder<'db> {
 
         let ty = match expression {
             ast::Expr::Name(name) => {
-                debug_assert!(
-                    name.ctx.is_load(),
-                    "name in a type expression is always 'load' but got: '{:?}'",
-                    name.ctx
-                );
-
+                debug_assert_eq!(name.ctx, ast::ExprContext::Load);
                 self.infer_name_expression(name).to_instance(self.db)
             }
 
+            ast::Expr::Attribute(attribute_expression) => {
+                debug_assert_eq!(attribute_expression.ctx, ast::ExprContext::Load);
+                self.infer_attribute_expression(attribute_expression)
```

Comment:
This is out of scope for this PR, but I think eventually when we are reusing value-expression methods like this in inferring a type expression, we will have to pass some context down making it explicit that we're in a type expression context, because there may be sub-expressions (like `Literal[1]`) that we would infer a different type for in a type-expression context.